pkg/volume/util/subpath/subpath_linux.go

```diff
@@ -241,7 +241,8 @@ func doCleanSubPaths(mounter mount.Interface, podDir string, volumeName string)
 			if err = doCleanSubPath(mounter, fullContainerDirPath, filepath.Base(path)); err != nil {
 				return err
 			}
-			return nil
+			// skip subdirs of the volume: it only matters the first level to unmount, otherwise it would try to unmount subdir of the volume
+			return filepath.SkipDir
```

Comment:
I'm not sure if this is going to work when the subpath could be a file. SkipDir will skip the rest of the directory in that case.  How about we change this back to use `ioutil.ReadDir`? I unfortunately can't remember why I had changed it to Walk in the first place.  Thanks for catching this!

Reply:
`ioutil.ReadDir` was my first try :) , but I've had some problems in some e2e tests with `file stale handle`. I guess `ioutil.ReadDir` is too eager

For files I'm expecting them to be in the first level and get unmounted
```
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/0 < my-dir
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/1 < my-file
``` 

But I'll add a test case to make it sure, I'll send an update here soon

Reply:
about the test case, just checked and it is already covered https://github.com/kubernetes/kubernetes/pull/82698/files#r341848461

Reply:
The existing case doesn't quite cover the scenario I'm thinking about.  According to golang docs, "If the function returns SkipDir when invoked on a non-directory file, Walk skips the remaining files in the containing directory."

Which I think is not what we want?  If "0" was a file and "1" was a directory, then I believe that means "1" will get skipped.

Reply:
@msau42 
very good point :+1: 

added case with files and directories

https://github.com/kubernetes/kubernetes/pull/82698/files#diff-b962320d780ac74dffce9da68eda4f4cR551

```
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/0 < my-file0
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/1 < my-dir1
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/2 < my-dir2
/var/lib/kubelet/pods/<uid>/volume-subpaths/<volume>/<container name>/3 < my-file3
```


let me know what you think about it